### PR TITLE
Implementing Google One Tap

### PR DESCRIPTION
**Pull request by indeewari**
Related to https://github.com/wso2/product-is/issues/14407
### Proposed changes in this pull request

Integrating Google One Tap - BE changes.
Generalizing OpenIDConnectAuthenticator to accommodate Google One Tap flow.

### When should this PR be merged

Dependencies
identity-framework
identity-apps

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 1]
```diff
@@ -452,95 +461,138 @@ private String getOIDCAuthzEndpoint(Map<String, String> authenticatorProperties)
     protected void processAuthenticationResponse(HttpServletRequest request, HttpServletResponse response,
                                                  AuthenticationContext context) throws AuthenticationFailedException {
 
-        try {
+        // oAuthResponse can be null in some authentication flows. i.e Google One Tap.
+        OAuthClientResponse oAuthResponse = generateOauthResponse(request, context);
+        // TODO : return access token and id token to framework
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java:466]: Is this under development?

Comment (indeewari) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java:466]: Early comment, shown as edited due to try catch block

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 2]
```diff
@@ -452,95 +461,138 @@ private String getOIDCAuthzEndpoint(Map<String, String> authenticatorProperties)
...
+        mapAccessToken(request, context, oAuthResponse);
+        String idToken = mapIdToken(context, request, oAuthResponse);
 
-            OAuthAuthzResponse authzResponse = OAuthAuthzResponse.oauthCodeAuthzResponse(request);
-            OAuthClientRequest accessTokenRequest = getAccessTokenRequest(context, authzResponse);
+        Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+        if (StringUtils.isBlank(idToken) && requiredIDToken(authenticatorProperties)) {
```
[the remark this answers is not in the document]
Comment (indeewari) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: Fixed 2bb7bf448d50cd3c75c2af84639f79103c292bbd

**Review by mevan-karu — APPROVED** [review]

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 3]
```diff
@@ -452,95 +461,138 @@ private String getOIDCAuthzEndpoint(Map<String, String> authenticatorProperties)
...
+     * Generates OAuth client and returns the oAuthResponse according to the flow supported by the authenticator.
+     * Overridden in Google Authenticator for Google one tap.
+     *
+     * @param request HttpServletRequest.
+     * @param context AuthenticationContext.
+     * @return OAuthClientResponse.
+     * @throws AuthenticationFailedException throws error when OAuthAuthzResponse validation fails for either error
+     *                                       response or the parameters.
+     */
+    protected OAuthClientResponse generateOauthResponse(HttpServletRequest request, AuthenticationContext context)
```
Comment (mevan-karu) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java:577]: `requestAccessToken` WDYT?

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 4]
```diff
@@ -452,95 +461,138 @@ private String getOIDCAuthzEndpoint(Map<String, String> authenticatorProperties)
...
-                    // Add 'sid' claim into authentication context, to be stored in the UserSessionStore for
-                    // single logout.
-                    context.setProperty(FEDERATED_IDP_SESSION_ID + idpName, sidClaim);
-                }
+    /**
+     * Retrieves or maps the ID token according to the flow supported by the authenticator.
+     * Overridden in Google Authenticator for Google one tap.
+     *
+     * @param context       AuthenticationContext.
+     * @param request       HttpServletRequest
```
Comment (lashinijay) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java:535]: missing full stops.

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 5]
```diff
@@ -452,95 +461,138 @@ private String getOIDCAuthzEndpoint(Map<String, String> authenticatorProperties)
...
             }
 
-            OIDCStateInfo stateInfoOIDC = new OIDCStateInfo();
-            stateInfoOIDC.setIdTokenHint(idToken);
-            context.setStateInfo(stateInfoOIDC);
-
-            context.setProperty(OIDCAuthenticatorConstants.ACCESS_TOKEN, accessToken);
+            if (log.isDebugEnabled() && IdentityUtil
+                    .isTokenLoggable(IdentityConstants.IdentityTokens.USER_ID_TOKEN)) {
+                log.debug("Retrieved the User Information:" + jwtAttributeMap);
```
Comment (lashinijay) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java:506]: Shall we change "log" to "LOG"?

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java (hunk 3 above)
Comment (indeewari) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: This is fixed 17aec3bd27b4fb46de26ee69631363ceabb32f2c

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java (hunk 4 above)
Comment (indeewari) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: This is fixed 17aec3bd27b4fb46de26ee69631363ceabb32f2c

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java (hunk 5 above)
Comment (indeewari) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: This is fixed

**Review by lashinijay — APPROVED** [review]

**Pull request merged**